ArmPkg/Drivers/SmmuDxe/SmmuV3Util.c

```diff
@@ -0,0 +1,649 @@
+  do {
+    RegVal = MmioRead32 (SmmuReg);
+    if ((RegVal & Mask) == Value) {
+      return EFI_SUCCESS;
+    }
+
+    MicroSecondDelay (100);
+  } while ((--Count) > 0);
+
+  DEBUG ((
+    DEBUG_ERROR,
+    "SmmuV3Poll: Timeout polling SMMUv3 register @%p Read value 0x%x "
+    "expected 0x%x\n",
+    SmmuReg,
+    RegVal,
+    ((Value == 0) ? (RegVal & ~Mask) : (RegVal | Mask))
+    ));
+
+  return EFI_TIMEOUT;
+}
+
+/**
+  Consume the event queue for errors and retrieve the fault record.
+
+  @param [in]  SmmuInfo     Pointer to the SMMU_INFO structure.
+  @param [out] FaultRecord  Pointer to the fault record structure.
+
+  @retval EFI_SUCCESS       Success.
+  @retval Others            Failure.
+**/
+EFI_STATUS
+EFIAPI
+SmmuV3ConsumeEventQueueForErrors (
+  IN SMMU_INFO             *SmmuInfo,
+  OUT SMMUV3_FAULT_RECORD  *FaultRecord
+  )
+{
+  SMMUV3_EVENTQ_CONS   Consumer;
+  UINT32               ConsumerIndex;
+  UINT32               ConsumerWrap;
+  SMMUV3_FAULT_RECORD  *NextFault;
+  SMMUV3_EVENTQ_PROD   Producer;
+  UINT32               ProducerIndex;
+  UINT32               ProducerWrap;
+  BOOLEAN              QueueEmpty;
+  UINT32               QueueMask;
+  UINT32               TotalQueueEntries;
+  UINT32               WrapMask;
+
+  TotalQueueEntries = SMMUV3_COUNT_FROM_LOG2 (SmmuInfo->EventQueueLog2Size);
+  WrapMask          = TotalQueueEntries;
+  QueueMask         = TotalQueueEntries - 1;
+
+  Producer.AsUINT32 = SmmuV3ReadRegister32 (SmmuInfo->SmmuBase + 0x10000, SMMU_EVENTQ_PROD);
+  Consumer.AsUINT32 = SmmuV3ReadRegister32 (SmmuInfo->SmmuBase + 0x10000, SMMU_EVENTQ_CONS);
+
+  ProducerIndex = Producer.WriteIndex & QueueMask;
+  ProducerWrap  = Producer.WriteIndex & WrapMask;
+  ConsumerIndex = Consumer.ReadIndex & QueueMask;
+  ConsumerWrap  = Consumer.ReadIndex & WrapMask;
+  QueueEmpty    = SMMUV3_IS_QUEUE_EMPTY (
+                    ProducerIndex,
+                    ProducerWrap,
+                    ConsumerIndex,
+                    ConsumerWrap
+                    );
+
+  if (QueueEmpty != FALSE) {
+    goto End;
+  }
+
+  NextFault = SmmuInfo->EventQueue + ConsumerIndex;
+  CopyMem (FaultRecord, NextFault, SMMUV3_EVENT_QUEUE_ENTRY_SIZE);
+
+  ConsumerIndex += 1;
+  if (ConsumerIndex == TotalQueueEntries) {
+    ConsumerIndex = 0;
+    ConsumerWrap  = ConsumerWrap ^ WrapMask;
+  }
+
+  Consumer.ReadIndex = ConsumerIndex | ConsumerWrap;
+
+  ArmDataSynchronizationBarrier ();
+
+  SmmuV3WriteRegister32 (SmmuInfo->SmmuBase + 0x10000, SMMU_EVENTQ_CONS, Consumer.AsUINT32);
+
+End:
+  return EFI_SUCCESS;
+}
+
+/**
+  Print the errors from the SMMUv3. Prints Event Queue entries and GError register.
+
+  @param [in]  SmmuInfo  Pointer to the SMMU_INFO structure.
+**/
+VOID
+EFIAPI
+SmmuV3PrintErrors (
+  IN SMMU_INFO  *SmmuInfo
+  )
+{
+  SMMUV3_GERROR        GError;
+  SMMUV3_FAULT_RECORD  FaultRecord = { 0 };
+
+  SmmuV3ConsumeEventQueueForErrors (SmmuInfo, &FaultRecord);
+  DEBUG ((DEBUG_INFO, "FaultRecord:\n"));

```

Comment:
error level here and below?

Reply:
fixed, this function now only logs errors if errors are found as DEBUG_ERROR.